### PR TITLE
Add edgeconnect version logic

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Description

In order to create a edge connect deployment, we want to store the image name, tag and digest stored in the edgeconnect status. This PR is about the handling of this version status field.

## How can this be tested?

- Setup edgeconnect CR 
- The version status field should change accordingly

## Checklist

- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly with a single label
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: src/controllers/csi/provisioner/controller.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ import (
 	"github.com/Dynatrace/dynatrace-operator/src/kubeobjects"
 	"github.com/pkg/errors"
 	"github.com/spf13/afero"
+	v1 "k8s.io/api/core/v1"
```
Comment (0sewa0) [inline src/controllers/csi/provisioner/controller.go:38]:
```suggestion
	corev1 "k8s.io/api/core/v1"
```

File: src/controllers/csi/provisioner/install.go [hunk 2]
```diff
@@ -13,6 +13,8 @@ import (
 	"github.com/Dynatrace/dynatrace-operator/src/installer/url"
 	"github.com/Dynatrace/dynatrace-operator/src/processmoduleconfig"
 	"github.com/spf13/afero"
+	v1 "k8s.io/api/core/v1"
```
Comment (0sewa0) [inline src/controllers/csi/provisioner/install.go:16]:
```suggestion
	corev1 "k8s.io/api/core/v1"
```

File: src/controllers/csi/provisioner/controller.go [hunk 3]
```diff
@@ -51,7 +52,7 @@ const (
 )
 
 type urlInstallerBuilder func(afero.Fs, dtclient.Client, *url.Properties) installer.Installer
-type imageInstallerBuilder func(afero.Fs, *image.Properties) installer.Installer
+type imageInstallerBuilder func(afero.Fs, *image.Properties, client.Reader, v1.Secret) (installer.Installer, error)
```
Comment (0sewa0) [inline src/controllers/csi/provisioner/controller.go:55]: nitpick:
```suggestion
type imageInstallerBuilder func(afero.Fs, client.Reader, *image.Properties, v1.Secret) (installer.Installer, error)
```

File: src/controllers/dynakube/version/image_hash.go [hunk 4]
```diff
@@ -11,7 +11,8 @@ import (
 	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/src/api/v1beta1/dynakube"
 	"github.com/Dynatrace/dynatrace-operator/src/dockerkeychain"
 	"github.com/Dynatrace/dynatrace-operator/src/registry"
-	"github.com/spf13/afero"
+	v1 "k8s.io/api/core/v1"
```
Comment (0sewa0) [inline src/controllers/dynakube/version/image_hash.go:14]:
```suggestion
	corev1 "k8s.io/api/core/v1"
```

File: src/controllers/dynakube/version/image_hash.go [hunk 5]
```diff
@@ -45,8 +46,17 @@ func GetImageVersion( //nolint:revive // argument-limit
 	var err error
 	var proxy string
 
-	keychain := dockerkeychain.NewDockerKeychain(registryAuthPath, afero.NewOsFs())
 	transport := http.DefaultTransport.(*http.Transport).Clone()
+	keychain := dockerkeychain.NewDockerKeychain()
+	err = keychain.LoadDockerConfigFromSecret(ctx, apiReader, v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dynakube.PullSecret(),
```
Comment (0sewa0) [inline src/controllers/dynakube/version/image_hash.go:53]: you could do some refactoring to make you life easier:
`dynakube.PullSecret() -> dynakube.PullSecretName()`

then create a `dynakube.PullSecret()`(with some other name that tells the reader that its an incomplete secret as it does not have the data part of it) that actually returns a secret, so we don't do this all the time.
- or you can make the `dynakube.PullSecret()` to return the actual pull secret (make the call and all), similar how the `dynakube.Proxy()` will make a call to get the proxy

File: src/controllers/edgeconnect/controller.go [hunk 6]
```diff
@@ -66,6 +67,13 @@ func (controller *Controller) Reconcile(ctx context.Context, request reconcile.R
 		return reconcile.Result{RequeueAfter: errorUpdateInterval}, nil
 	}
 
+	log.Info("updating version info", "name", request.Name, "namespace", request.Namespace)
+	versionReconciler := version.NewReconciler(edgeConnect, controller.apiReader, timeprovider.New().Freeze())
+	if err = versionReconciler.Reconcile(ctx); err != nil {
```
Comment (0sewa0) [inline src/controllers/edgeconnect/controller.go:72]: nitpick: `timeprovider.Freeze()` is mainly meant for testing, I don't think it causes problems here, but I also see no benefit. 🤷

File: src/controllers/edgeconnect/version/edgeConnect_test.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/opencontainers/go-digest"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const fakeDigest = "sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"
+
+type registryClientMock struct {
```
Comment (0sewa0) [inline src/controllers/edgeconnect/version/edgeConnect_test.go:24]: a mock for this already exists: https://github.com/Dynatrace/dynatrace-operator/blob/c1b3de905ad44a3e82b4204ac6806ce862b51b5d/src/registry/mocks/client.go#L15

(you would still need to set it up)

File: src/controllers/edgeconnect/version/edgeconnect.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/Dynatrace/dynatrace-operator/src/timeprovider"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const DefaultMinRequestThreshold = 15 * time.Minute
```
Comment (0sewa0) [inline src/controllers/edgeconnect/version/edgeconnect.go:23]: why is this here defined?
for the dynakube its defined [here](https://github.com/Dynatrace/dynatrace-operator/blob/c1b3de905ad44a3e82b4204ac6806ce862b51b5d/src/api/v1beta1/dynakube/feature_flags.go#L121)

File: src/dockerkeychain/docker_keychain.go [hunk 9]
```diff
@@ -1,53 +1,84 @@
...
-func NewDockerKeychain(dockerconfigFile string, filesystem afero.Fs) authn.Keychain {
-	return &dockerKeychain{
-		dockerConfigFile: dockerconfigFile,
-		filesystem:       filesystem,
+func NewDockerKeychain() *DockerKeychain {
+	keychain := &DockerKeychain{}
+	return keychain
+}
+
+func (keychain *DockerKeychain) LoadDockerConfigFromSecret(ctx context.Context, apiReader client.Reader, pullSecret corev1.Secret) error {
```
Comment (0sewa0) [inline src/dockerkeychain/docker_keychain.go:28]: is there a benefit to not just do this in the `NewDockerKeychain` ?

File: src/installer/image/installer.go [hunk 10]
```diff
@@ -44,15 +46,15 @@ func GetDigest(uri string) (string, error) {
 	return canonRef.Digest().Encoded(), nil
 }
 
-func NewImageInstaller(fs afero.Fs, props *Properties) installer.Installer {
+func NewImageInstaller(fs afero.Fs, props *Properties, apiReader client.Reader, pullSecret v1.Secret) (installer.Installer, error) {
```
Comment (0sewa0) [inline src/installer/image/installer.go:49]:
```suggestion
func NewImageInstaller(fs afero.Fs, apiReader client.Reader, props *Properties, pullSecret v1.Secret) (installer.Installer, error) {
```

(also `Properties` already has an `ApiReader`)

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: src/dockerkeychain/docker_keychain.go (hunk 9 above)
Comment (andriisoldatenko) [reply src/dockerkeychain/docker_keychain.go:28]: I agree with @0sewa0 ☝️

File: src/installer/image/installer.go (hunk 10 above)
Comment (andriisoldatenko) [reply src/installer/image/installer.go:49]: @0sewa0 I saw you mentioned it in another PRs: let's add it to CONTRIBUTING.md -> code-styles

File: src/controllers/dynakube/version/image_hash.go (hunk 5 above)
Comment (gkrenn) [reply src/controllers/dynakube/version/image_hash.go:53]: Will create a follow up ticket to tackle this

Comment (gkrenn) [reply src/controllers/dynakube/version/image_hash.go:53]: Or not, moved the `PullSecret` function to `PullSecretName` and created a new function called `PullSecretWithoutData`

File: src/controllers/edgeconnect/version/edgeConnect_test.go (hunk 7 above)
Comment (gkrenn) [reply src/controllers/edgeconnect/version/edgeConnect_test.go:24]: I did this as it is easier to read if just the interface is implemented and one can immediately see what the getimageversion is doing

File: src/controllers/edgeconnect/version/edgeconnect.go (hunk 8 above)
Comment (gkrenn) [reply src/controllers/edgeconnect/version/edgeconnect.go:23]: put it into a own config.go file in the edgeconnect package

File: src/dockerkeychain/docker_keychain.go (hunk 9 above)
Comment (gkrenn) [reply src/dockerkeychain/docker_keychain.go:28]: I did this for unit tests, but it is no longer required, so will change it back

File: src/installer/image/installer.go (hunk 10 above)
Comment (0sewa0) [reply src/installer/image/installer.go]: created PR: https://github.com/Dynatrace/dynatrace-operator/pull/2058

File: src/controllers/edgeconnect/version/edgeConnect_test.go (hunk 7 above)
Comment (0sewa0) [reply src/controllers/edgeconnect/version/edgeConnect_test.go:24]: it would be the same level of readable (just create a "setupMockedGetter", where you setup the mock), but you would have more tools to check what is happening, and the actual mock would catch unexpected calls.

but, as its a simple interface I will consider this a nitpick :D

File: src/controllers/csi/provisioner/install.go [hunk 11]
```diff
@@ -35,16 +37,20 @@ func (provisioner *OneAgentProvisioner) installAgentImage(ctx context.Context, d
 		return "", err
 	}
 
-	imageInstaller := provisioner.imageInstallerBuilder(provisioner.fs, &image.Properties{
+	pullSecret := corev1.Secret{
```
Comment (0sewa0) [inline src/controllers/csi/provisioner/install.go:40]: use the function you created :D

File: src/api/v1alpha1/edgeconnect/properties.go [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+	}
+
+	return fmt.Sprintf("%s:%s", repository, tag)
+}
+
+func (edgeConnect *EdgeConnect) IsCustomImage() bool {
+	return edgeConnect.Spec.ImageRef.Repository != ""
+}
+
+func (edgeConnect *EdgeConnect) PullSecretWithoutData() corev1.Secret {
```
Comment (0sewa0) [inline src/api/v1alpha1/edgeconnect/properties.go:31]: nitpick: add a comment so we know what it is meant to be used for ("why does method exist")

File: src/installer/image/installer.go [hunk 13]
```diff
@@ -44,15 +46,15 @@ func GetDigest(uri string) (string, error) {
 	return canonRef.Digest().Encoded(), nil
 }
 
-func NewImageInstaller(fs afero.Fs, props *Properties) installer.Installer {
+func NewImageInstaller(fs afero.Fs, props *Properties, pullSecret corev1.Secret) (installer.Installer, error) {
```
Comment (0sewa0) [inline src/installer/image/installer.go:49]: bit misleading, someone might think that the `pullSecret` is actually a complete secret, but its created with `dynakube.PullSecretWithoutData`

you already have `Dynakube` in the properties, so you don't need to pass this

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

**Review by 0sewa0 — APPROVED** [review]
